[BEAM-124] Flink and Spark running Examples WordCountIT

jasonkuster opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

jasonkuster commented:
R: @mxm @amitsela 

Hey Max, Amit.

Looking for some feedback on this pull request. The purpose is to remove the dependencies of Spark and Flink runner on Beam to enable them to run the WordCountIT in examples, as Dataflow currently does. As things were in the codebase, both Spark and Flink depended on examples for some of the code in WordCount.java and TfIdf.java. In the Flink case I've removed the tests; in the Spark case I've just added the code in. I'd love to hear your guys' thoughts on what the right thing to do is going forward.

The benefit we get from this is that this is where the new End-to-End tests seem to be going, such that they can be written in a runner-agnostic way and then run just by flipping a few flags (for example, see the commands below for running this test). Let me know your thoughts!

mvn clean verify -pl examples/java -am -rf :java-examples-all -DskipITs=false -DintegrationTestPipelineOptions='[ "--tempRoot=/tmp", "--inputFile=/tmp/kinglear.txt", "--runner=org.apache.beam.runners.spark.SparkPipelineRunner", "--sparkMaster=local" ]'

mvn clean verify -pl examples/java -am -rf :java-examples-all -DskipITs=false -DintegrationTestPipelineOptions='[ "--tempRoot=/tmp", "--inputFile=/tmp/kinglear.txt", "--runner=org.apache.beam.runners.flink.FlinkPipelineRunner" ]

mvn clean verify -pl examples/java -am -rf :java-examples-all -DskipITs=false -DintegrationTestPipelineOptions='[ "--tempRoot=gs://clouddfe-testing-temp-storage", "--runner=org.apache.beam.sdk.testing.TestDataflowPipelineRunner" ]'

Jason

aljoscha commented:
I think this is the right way to go. In #343 I'm also removing these two examples because all RunnableOnService tests will be executed on Flink with those changes.

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/io/NumShardsTest.java:
```diff
@@ -61,6 +68,66 @@
+   * Count) as a reusable PTransform subclass. Using composite transforms allows for easy reuse,
+   * modular testing, and an improved monitoring experience.
+   */
+  public static class CountWords extends PTransform<PCollection<String>,
```
Could use `SimpleWordCountTest.CountWords` instead. Maybe need to make a small change to the format function.

jasonkuster replied:
I'll take a look. Thanks!

amitsela commented on examples/java/pom.xml:
```diff
@@ -220,6 +220,30 @@
+      <version>${project.version}</version>
+    </dependency>
+
+    <dependency>
```
I guess the Spark dependencies are here because they are provided in the runner and therefore they are not transitive. I'm afraid this will cause the examples pom to be a bit "busy" as more runners join...

davorbonaci replied:
At a minimum, these should be optional, runtime deps only.

amitsela commented:
I generally agree with @aljoscha and once https://github.com/apache/incubator-beam/pull/294 is done, and RunnableOnService tests will cover those use cases, they might be removed from Spark runner tests as well. 
@davorbonaci your thoughts on the examples pom.xml ?

mxm commented:
Hi @jasonkuster! +1 for enabling end-to-end tests for all Runners. A couple questions: I wonder why do you remove a Flink TfIdf integration test and add one for Spark? ☺️ Presumably because the `RunnableOnService` tests are not yet integrated with the Spark Runner?

jasonkuster commented:
Hey @mxm! I removed in Flink and added in Spark just to see what the two different methods of resolving the dependency issues would look like. I'm happy to do either for either one, but based on the above comments it looks like the `RunnableOnService` tests are in process on both Spark and Flink, so once those are done and in it sounds like the right thing to do is just to remove the offending tests. I'm flexible though - my goal is just to get the E2E tests running everywhere. :smile:

davorbonaci commented:
(should be rebased, given relevant changes to the pom.)

mxm commented:
@jasonkuster The `RunnableOnService` tests are integrated with the Flink Runner for batched execution. So removing batch examples is fine. The streaming side still needs side inputs to support the tests.

+1 for merging from my side (needs rebasing though)
